### PR TITLE
coord: Fix builtin schema migration create order

### PR DESCRIPTION
**Pull request by jkosh44**
It is important for built in schema migrations to recreate all objects
in the new dependency order, that is parents before children.
Previously, we failed to do that for the following reasons.
- The visit order of BFS is not guaranteed to be sorted in topological
order.
- The dependencies between builtin objects may change after a restart.
The previous code was using the persisted, possibly inaccurate,
dependencies.

This commit fixes both of these issues so that objects are recreated in
the correct order.

### Motivation
This PR fixes a previously unreported bug.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [X] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-protobuf` label.
- [X] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - N/A

### CODE REVIEW

**Review by mjibson — APPROVED** [review]
Your idea to add some unit tests to this sounds real good. Can you add some small ones that would have caught this problem?

**Comment by jkosh44** [comment]
After chatting with Frank, I've realized that the delete order is also wrong. It is not sorted in reverse topological order because the visit order of BFS is not guaranteed to be sorted in topological order.

**Comment by jkosh44** [comment]
This should be fixed now, but I still need to add some tests which will be tough. I also updated the `fingerprint` function so it's less aggressive.

File: src/adapter/src/catalog/builtin.rs [hunk 1]
```diff
@@ -162,16 +163,28 @@ impl Fingerprint for BuiltinFunc {
...
     }
 }
 
+impl Fingerprint for RelationDesc {
+    fn fingerprint(&self) -> u64 {
+        self.typ().fingerprint()
+    }
+}
+
+impl Fingerprint for RelationType {
```
Comment (jkosh44) [inline src/adapter/src/catalog/builtin.rs:182]: This is still probably even more aggressive than we ultimately want. Somewhere nested in `RelationType` is `ScalarType`. When a new type is added, the derived hash implementation changes and will change the fingerprint for all objects that use this impl.

**Comment by jkosh44** [comment]
@mjibson It took awhile, but I finally updated this. 378 of the new lines are tests, so don't be too scared by the total new lines. When you get a chance can you please take another look?

**Comment by jkosh44** [comment]
This isn't a high priority though, since it's a breaking change and likely won't make it into the release today. So it will have to wait until the final breaking change release.

File: src/adapter/src/catalog.rs [hunk 2]
```diff
@@ -2768,6 +2783,16 @@ impl<S: Append> Catalog<S> {
         self.storage().await.allocate_user_id().await
     }
 
+    pub async fn test_only_dont_use_in_production_allocate_system_id(
```
Comment (mjibson) [inline src/adapter/src/catalog.rs:2786]: `#[cfg(test)]` might work here? I don't know the rust details if it'll be compiled in for things in the `tests` directory or not, but worth a try. :shrug:

File: src/adapter/src/catalog/builtin.rs [hunk 3]
```diff
@@ -137,6 +137,7 @@ pub struct BuiltinRole {
     pub name: &'static str,
 }
 
+/// Uniquely identifies the definition of a builtin object.
```
Comment (mjibson) [inline src/adapter/src/catalog/builtin.rs:140]: What happens if a two different objects have the same fingerprint? Do fingerprints have to be unique among all other fingerprints, or just ones of the same object type? The word "uniquely" here conflicts with the various implementations that sometimes all produce 0, and also doesn't work with Hashable which may collide.

**Review by mjibson — APPROVED** [review]

**Comment by jkosh44** [comment]
@mjibson I really want to get this in the release because it's a breaking change and the release was pushed back. I'm going to merge this as is and fix make the changes in a follow up.

File: src/adapter/src/catalog/builtin.rs (hunk 3 above)
Comment (jkosh44) [reply src/adapter/src/catalog/builtin.rs:140]: Yeah this comment is incorrect. Fingerprints are never compared across objects. The only thing we care about is that when an object definition is changed, then the fingerprint changes.

Comment (mjibson) [reply src/adapter/src/catalog/builtin.rs:140]: Why is it ok to use hashed for that?

Comment (jkosh44) [reply src/adapter/src/catalog/builtin.rs:140]: Maybe I'm misunderstanding something, but if we change the object definition, then it's hash will also change.

Comment (jkosh44) [reply src/adapter/src/catalog/builtin.rs:140]: Maybe we should just be serializing the definition and comparing those.

Comment (mjibson) [reply src/adapter/src/catalog/builtin.rs:140]: I think it's hash *probably* changes, not will for sure change.

Comment (jkosh44) [reply src/adapter/src/catalog/builtin.rs:140]: hmm, yeah. I think a serialized json of the definition is probably what we want here. At least this is an existing issue so it should be OK to include this in the release.

**Pull request merged**